services/provider/server/consumer.go

```diff
@@ -0,0 +1,158 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ocsConsumerManager struct {
+	client      client.Client
+	ticketStore map[string]string
+	uidStore    map[string]string
+	mutex       sync.RWMutex
+}
+
+func NewConsumerManager(ctx context.Context, client client.Client) (*ocsConsumerManager, error) {
+	consumers := &ocsv1alpha1.StorageConsumerList{}
+	err := client.List(ctx, consumers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list storage consumers. %v", err)
+	}
+
+	ticketStore := map[string]string{}
+	uidStore := map[string]string{}
+
+	for _, consumer := range consumers.Items {
+		uid := string(consumer.UID)
+		if _, ok := uidStore[uid]; !ok {
+			uidStore[uid] = consumer.Name
+		}
+
+		if ticket, ok := consumer.GetAnnotations()[TicketAnnotation]; ok {
+			if _, ok := ticketStore[ticket]; !ok {
+				ticketStore[ticket] = consumer.Name
+			}
+		}
+	}
+
+	return &ocsConsumerManager{
+		client:      client,
+		ticketStore: ticketStore,
+		uidStore:    uidStore,
+	}, nil
+}
+
+// Create creates a new storageConsumer resource, updates the consumer cache and returns the storageConsumer UUID
+func (c *ocsConsumerManager) Create(ctx context.Context, name, namespace, ticket string, capacity resource.Quantity) (string, error) {
+	consumerObj := &ocsv1alpha1.StorageConsumer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Annotations: map[string]string{
+				TicketAnnotation: ticket,
+			},
+		},
+		Spec: ocsv1alpha1.StorageConsumerSpec{
+			Capacity: capacity,
+		},
+	}
+
+	err := c.client.Create(ctx, consumerObj)
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			return "", err
+		}
+		return "", fmt.Errorf("failed to create storageConsumer resource %q. %v", consumerObj.Name, err)
+	}
+
+	existingConsumer := &ocsv1alpha1.StorageConsumer{}
+	err = c.client.Get(ctx, types.NamespacedName{Name: consumerObj.Name, Namespace: consumerObj.Namespace}, existingConsumer)
+	if err != nil {
+		return "", fmt.Errorf("failed to get UUID from the storageConsumer resource %q. %v", consumerObj.Name, err)
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	consumerID := string(existingConsumer.UID)
+	if _, ok := c.uidStore[consumerID]; !ok {
+		c.uidStore[consumerID] = existingConsumer.Name
+	}
+	if _, ok := c.ticketStore[ticket]; !ok {
+		c.ticketStore[ticket] = existingConsumer.Name
+	}
+
+	return consumerID, nil
+}
+
+// Delete deletes the storageConsumer resource using UID and updates the consumer cache
+func (c *ocsConsumerManager) Delete(ctx context.Context, namespace, id string) error {
+	c.mutex.RLock()
+	consumerName, ok := c.uidStore[id]
+	if !ok {
+		klog.Warningf("no storageConsumer found with UUID %q", id)
+		c.mutex.RUnlock()
+		return nil
+	}
+	c.mutex.RUnlock()
```

Comment:
Why do we need to lock on get operations?

Reply:
discussed on the gchat that we need lock for read operations.